Hazel/Udp/UdpConnection.Fragmented.cs

```diff
@@ -0,0 +1,241 @@
+
+                buffer[0] = (byte)UdpSendOption.Fragment;
+
+                AttachReliableID(buffer, 1, () =>
+                {
+                    acksReceived++;
+
+                    if (acksReceived >= fragmentsCount)
+                    {
+                        ackCallback?.Invoke();
+                    }
+                });
+
+                buffer[3] = (byte)fragmentsCount;
+                buffer[4] = (byte)(fragmentsCount >> 8);
+
+                buffer[5] = (byte)id;
+                buffer[6] = (byte)(id >> 8);
+
+                var includingHeader = i == 0 && includeHeader;
+                if (includingHeader)
+                {
+                    buffer[7] = sendOption;
+                }
+
+                Buffer.BlockCopy(data, fragmentDataSize * i - (includingHeader ? 0 : 1), buffer, FragmentHeaderSize + (includingHeader ? 1 : 0), dataLength - (includingHeader ? 1 : 0));
+
+                WriteBytesToConnection(buffer, buffer.Length);
+            }
+        }
+
+        protected void FragmentMessageReceive(MessageReader messageReader)
+        {
+            if (ProcessReliableReceive(messageReader.Buffer, 1, out var id))
+            {
+                messageReader.Position += 3;
+
+                var fragmentsCount = messageReader.ReadUInt16();
+                var fragmentedMessageId = messageReader.ReadUInt16();
+
+                lock (_fragmentedMessagesReceived)
+                {
+                    if (!_fragmentedMessagesReceived.TryGetValue(fragmentedMessageId, out var fragmentedMessage))
+                    {
+                        _fragmentedMessagesReceived.Add(fragmentedMessageId, fragmentedMessage = new FragmentedMessage(fragmentsCount));
+                    }
+
+                    var buffer = new byte[messageReader.Length - messageReader.Position];
+                    Buffer.BlockCopy(messageReader.Buffer, messageReader.Position, buffer, 0, messageReader.Length - messageReader.Position);
+
+                    fragmentedMessage.Fragments.Add(new FragmentedMessage.Fragment(id, buffer));
+
+                    if (fragmentedMessage.Fragments.Count == fragmentsCount)
+                    {
+                        var reconstructed = fragmentedMessage.Reconstruct();
+                        InvokeDataReceived(SendOption.Reliable, MessageReader.Get(reconstructed), 1, reconstructed.Length);
+
+                        _fragmentedMessagesReceived.Remove(id);
+                    }
+                }
+            }
+        }
+
+        protected class FragmentedMessage
+        {
+            /// <summary>
+            ///     The total number of fragments expected.
+            /// </summary>
+            public int FragmentsCount { get; }
+
+            /// <summary>
+            ///     The fragments received so far.
+            /// </summary>
+            public HashSet<Fragment> Fragments { get; } = new HashSet<Fragment>();
```

Comment:
Since you know the number of fragments, you should use an array and index into it. It's way faster.